opa/opa_test.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/jarcoal/httpmock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("opa client", func() {
+	var (
+		Opa       Client
+		opaPolicy string
+	)
+
+	BeforeEach(func() {
+		Opa = Client{
+			logger: logger,
+			Host:   fmt.Sprintf("http://%s", gofakeit.DomainName()),
+		}
+		opaPolicy = gofakeit.Word()
+	})
+
+	When("a new OPA client is created", func() {
+		It("returns OpaClient", func() {
+			host := fmt.Sprintf("http://%s", gofakeit.DomainName())
+			opa := NewClient(logger, host)
+			Expect(opa).To(BeAssignableToTypeOf(&Client{}))
+			Expect(opa.Host).To((Equal(host)))
+		})
+	})
+
+	Context("an OPA policy is intialized", func() {
+		var (
+			initializePolicyError ClientError
+			getPolicyResponse     *http.Response
+			getPolicyError        error
+		)
+
+		BeforeEach(func() {
+			getPolicyResponse = httpmock.NewStringResponse(200, "{}")
+			getPolicyError = nil
+
+			httpmock.RegisterResponder("GET", fmt.Sprintf("%s/v1/policies/%s", Opa.Host, opaPolicy),
+				func(req *http.Request) (*http.Response, error) {
+					return getPolicyResponse, getPolicyError
+				},
+			)
+		})
+
+		JustBeforeEach(func() {
+			initializePolicyError = Opa.InitializePolicy(opaPolicy)
+		})
+
+		It("should check if policy exists in OPA", func() {
+			Expect(httpmock.GetTotalCallCount()).To(Equal(1))
+		})
+
+		When("fetch policy request returns error", func() {
+			BeforeEach(func() {
+				getPolicyResponse = nil
+				getPolicyError = errors.New("error connecting to host")
+			})
+
+			It("should return an http request error", func() {
+				Expect(initializePolicyError).To(HaveOccurred())
+				Expect(initializePolicyError.Type()).To(Equal(OpaClientErrorTypePolicyExists))
+				Expect(initializePolicyError.CausedBy()).To(HaveOccurred())
+				Expect(initializePolicyError.CausedBy()).To(BeAssignableToTypeOf(clientError{}))
+				Expect(initializePolicyError.CausedBy().(clientError).Type()).To(Equal(OpaClientErrorTypeHTTP))
+			})
+		})
+
+		When("fetch policy response status is not OK", func() {
+			BeforeEach(func() {
+				getPolicyResponse = httpmock.NewStringResponse(http.StatusInternalServerError, "OPA Error")
+			})
+
+			It("should return an bad response error", func() {
+				Expect(initializePolicyError).To(HaveOccurred())
+				Expect(initializePolicyError.Type()).To(Equal(OpaClientErrorTypePolicyExists))
+				Expect(initializePolicyError.CausedBy()).To(HaveOccurred())
+				Expect(initializePolicyError.CausedBy()).To(BeAssignableToTypeOf(clientError{}))
+				Expect(initializePolicyError.CausedBy().(clientError).Type()).To(Equal(OpaClientErrorTypeBadResponse))
+			})
+		})
+
+		When("policy exists", func() {
+			It("should do nothing", func() {
+				Expect(initializePolicyError).NotTo(HaveOccurred())
+			})
+		})
+
+		When("policy does not exist", func() {
+			BeforeEach(func() {
+				getPolicyResponse = httpmock.NewStringResponse(404, "{}")
+			})
+
+			It("should fetch policy violations from Elasticsearch", func() {
```

Comment:
is this needed?

Reply:
its a placeholder for functionality what hasn't been added yet